contracts/0.8.25/vaults/VaultHub.sol

```diff
@@ -4,376 +4,294 @@
-        uint256 reserveRatioBP = socket.reserveRatioBP;
-        uint256 maxMintableRatio = (TOTAL_BASIS_POINTS - reserveRatioBP);
-        uint256 totalValue = IStakingVault(_vault).totalValue();
+    /// @return amount to rebalance or UINT256_MAX if it's impossible to make the vault healthy using rebalance
+    /// @dev returns 0 if the vault is not connected
+    function rebalanceShortfall(address _vault) external view returns (uint256) {
+        Storage storage $ = _storage();
+        return _rebalanceShortfall(
+            $.connections[_vault],
+            $.records[_vault]
+        );
+    }
 
-        // Impossible to rebalance a vault with deficit
-        if (liabilityStETH >= totalValue) {
-            // return MAX_UINT_256
-            return type(uint256).max;
-        }
+    /// @notice Set if a vault proxy codehash is allowed to be connected to the hub
+    /// @param _codehash vault proxy codehash
+    /// @param _allowed true to add, false to remove
+    function setAllowedCodehash(bytes32 _codehash, bool _allowed) external onlyRole(VAULT_CODEHASH_SET_ROLE) {
+        if (_codehash == bytes32(0)) revert ZeroArgument();
+        if (_codehash == EMPTY_CODEHASH) revert ZeroCodehash();
 
-        // (liabilityStETH - X) / (vault.totalValue() - X) = maxMintableRatio / TOTAL_BASIS_POINTS
-        // (liabilityStETH - X) * TOTAL_BASIS_POINTS = (vault.totalValue() - X) * maxMintableRatio
-        // liabilityStETH * TOTAL_BASIS_POINTS - X * TOTAL_BASIS_POINTS = totalValue * maxMintableRatio - X * maxMintableRatio
-        // X * maxMintableRatio - X * TOTAL_BASIS_POINTS = totalValue * maxMintableRatio - liabilityStETH * TOTAL_BASIS_POINTS
-        // X * (maxMintableRatio - TOTAL_BASIS_POINTS) = vault.totalValue() * maxMintableRatio - liabilityStETH * TOTAL_BASIS_POINTS
-        // X = (vault.totalValue() * maxMintableRatio - liabilityStETH * TOTAL_BASIS_POINTS) / (maxMintableRatio - TOTAL_BASIS_POINTS)
-        // X = (liabilityStETH * TOTAL_BASIS_POINTS - vault.totalValue() * maxMintableRatio) / (TOTAL_BASIS_POINTS - maxMintableRatio)
-        // reserveRatio = TOTAL_BASIS_POINTS - maxMintableRatio
-        // X = (liabilityStETH * TOTAL_BASIS_POINTS - totalValue * maxMintableRatio) / reserveRatio
+        Storage storage $ = _storage();
+        $.codehashes[_codehash] = _allowed;
 
-        return (liabilityStETH * TOTAL_BASIS_POINTS - totalValue * maxMintableRatio) / reserveRatioBP;
+        emit AllowedCodehashUpdated(_codehash, _allowed);
     }
 
     /// @notice connects a vault to the hub in permissionless way, get limits from the Operator Grid
     /// @param _vault vault address
-    function connectVault(address _vault) external {
+    /// @dev vault should have transferred ownership to the VaultHub contract
+    function connectVault(address _vault) external whenResumed {
+        if (_vault == address(0)) revert VaultZeroAddress();
+
+        IStakingVault vault_ = IStakingVault(_vault);
```

Comment:
I'd say we shall check the `CONNECT_DEPOSIT` here

Reply:
I'd rather stay as it is, because CONNECT_DEPOSIT is tightly coupled with internal connection logic and the check looks natural there.